Feature/parser vseti

EkaterinaTerentyeva opened the pull request.

EkaterinaTerentyeva committed:
First commit

EkaterinaTerentyeva committed:
Corrected annotation types get_description_date

EkaterinaTerentyeva committed:
Changed the path to csv file

PhySci commented on src/parser/vseti.py:
```diff
@@ -0,0 +1,174 @@
+    return driver
+
+
+driver = webdriver.Chrome()
```
Может быть занести эту инициализацию драйвера в функцию main? Точно ли надо это делать здесь?

PhySci commented on src/parser/vseti.py:
```diff
@@ -0,0 +1,174 @@
+driver = webdriver.Chrome()
+
+# for Linux
+# driver = set_driver()
```
Можно вот так определять  ОС и решать, запускать инициализацию драйвера или нет
https://stackoverflow.com/questions/1854/how-to-identify-which-os-python-is-running-on

PhySci commented on src/parser/vseti.py:
```diff
@@ -0,0 +1,174 @@
+
+
+def get_description_date(link: str) -> Tuple[str, Union[date, None]]:
+    """opens a link for each vacancy and retrieves information about the vacancy description
```
Лайк за подробные комментарии.

PhySci commented on src/parser/vseti.py:
```diff
@@ -0,0 +1,174 @@
+        else:
+            date_publication = None
+        return description, date_publication
+    except requests.exceptions.RequestException as e:
```
Может быть поднять except до 64 строки, чтобы он действовал только на requests.get(). Тогда станет чуть понятнее, что именно ловит этот try.

PhySci commented on src/parser/vseti.py:
```diff
@@ -0,0 +1,174 @@
+        element = soup.find(tag, attrs)
+    else:
+        element = soup.find(tag)
+    if element:
```
Вот здесь у тебя два вложенных if - это увеличивает сложность понимания. Хорошо бы применить "ранный выход", т.е. ты проверяешь условие и если оно верно, то сразу выходишь, а иначе продолжаешь работу. Так ты избавляешь от "вложенной" логики.

```
if element is None:
    logging.info(...)
    return ''

if attribute:
    return element.get(attribute)
else:
    return element.text.strip()    
```

EkaterinaTerentyeva replied:
Здесь я не уверена что правильно сделала

PhySci commented on src/parser/vseti.py:
```diff
@@ -0,0 +1,174 @@
+    """returns information about vacancies with filtering for interns and juniors and given period of time,
+    which cannot be set in filters with selenium or directly in the url"""
+    period = date.today() - timedelta(days=int(args['days']))
+    try:
```
Два вложенных try-except - это прямо нехорошо, потому что сложно для понимания и могут как-то странно себя повести. Давай объединим их в один.
```
try:
    driver.get(url)
    driver.maximize_window()
    sleep(60)
    driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")
    sleep(60)
     jobs = driver.find_elements(By.XPATH, "//div[@class='collection-item w-dyn-item']")
except NoSuchElementException as e
  some code here
except WebDriverException as e
 some code here  
```

PhySci commented on src/parser/vseti.py:
```diff
@@ -0,0 +1,174 @@
+def scrape_job_cards(url: str) -> List[Tuple[str, str, str, str, str, str, str, date, str, str, str, str]]:
+    """returns information about vacancies with filtering for interns and juniors and given period of time,
+    which cannot be set in filters with selenium or directly in the url"""
+    period = date.today() - timedelta(days=int(args['days']))
```
А как переменная args попадает сюда?

PhySci commented on src/parser/vseti.py:
```diff
@@ -0,0 +1,174 @@
+    return vars(parser.parse_args())
+
+
+def main(args):
```
Давай вместо одной входной переменной args заведём две:
- filename
- days

и будем распаковывать эти переменные при вызове main. Так будет понятнее, какие именно аргументы принимает main
т.е.

```
def main(filename, days):
   ...
   
 if __name__ == '__main__':
    setup_logging("log.txt")
    args = parse_args()
    main(args["filename"], args["days"])  
```

PhySci reviewed (changes requested):
В целом, скрипт очень сеньёрный. Есть несколько минорных комментариев, которые хорошо бы поправить. Но они не влияют на качество кода, скорее про снижение когнитивной сложности. 
  Сам скрипт написан очень хорошо - выделены функции, имена переменных понятны, PEP8 присутствует. Отлично!

EkaterinaTerentyeva committed:
Function set_driver moved to utils.py
The choice of the platform added to def set_driver
Except moved closer to try in def get_description_date
Used early exit in def find_element_text
Removed second try in def scrape_job_cards
Added argument days_ago in def scrape_job_cards
Changed arguments in def main

EkaterinaTerentyeva committed:
Changed default number of days for the scraping

PhySci commented on src/parser/vseti.py:
```diff
@@ -0,0 +1,152 @@
+
+def find_element_text(soup: BeautifulSoup, tag: str, attrs: dict = None, attribute: str = None) -> str:
+    """Find element by tag and attributes, and get text or attribute by name"""
+    if attrs:
```
А может быть этот if совсем не нужен?
Если оставить только 
```
element = soup.find(tag, attrs) 
```
то при пустом attrs будет выполнять так, будто это
```
soup.find(tag) 
```

EkaterinaTerentyeva replied:
Я вот уже не помню почему, но почему то такой вариант у меня с треском грохался
Можно, конечно, вот так написать (исключительно для любителей красоты)
```
element = soup.find(tag, attrs) if attrs else soup.find(tag)
if not element:
        logging.info(f'bla-bla-bla}')
        return ''
return element.get(attribute) if attribute else element.text.strip()
```

PhySci commented:
Well done!
